modules/concurrency/src/spinner.cpp

```diff
@@ -64,34 +81,60 @@ void Spinner::spin() {
 
   try {
     start_timestamp_ = std::chrono::system_clock::now();
+    uint64_t spin_count_{ 0 };
 
     while (!stop_requested_.load()) {
-      const auto spin_result = stoppable_callback_();
-
-      ++spin_count_;
-
-      if (spin_result == SpinResult::STOP) {
+      state_ = attemptBinaryTransition(state_, { .input_state = State::NOT_INITIALIZED,
+                                                 .operation = callbacks_.init_cb,
+                                                 .success_state = State::INIT_SUCCESSFUL,
+                                                 .failure_state = State::INIT_FAILED });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::INIT_FAILED,
+                                                 .operation = callbacks_.shall_re_init_cb,
+                                                 .success_state = State::NOT_INITIALIZED,
+                                                 .failure_state = State::TERMINATE });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::READY_TO_SPIN,
+                                                 .operation = callbacks_.spin_once_cb,
+                                                 .success_state = State::SPIN_SUCCESSFUL,
+                                                 .failure_state = State::SPIN_FAILED });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::SPIN_FAILED,
+                                                 .operation = callbacks_.shall_re_init_cb,
+                                                 .success_state = State::NOT_INITIALIZED,
+                                                 .failure_state = State::TERMINATE });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::SPIN_SUCCESSFUL,
+                                                 .operation = callbacks_.shall_stop_,
+                                                 .success_state = State::TERMINATE,
+                                                 .failure_state = State::READY_TO_SPIN });
+
+      if (state_ == State::TERMINATE) {
         break;
       }
 
+      // Logic to handle periodic spinning at a fixed rate.
       if (spin_period_.count() == 0) {
         continue;
       }
-
-      const auto target_timestamp = start_timestamp_ + spin_count_ * spin_period_;
+      const auto target_timestamp = start_timestamp_ + (++spin_count_ * spin_period_);
```

Comment:
There is an issue here. This logic doesn't old true anymore as we also execute init code inside here. We need to update to keep track only of the time spent spinning.